conans/test/functional/generators/components/cmake_find_package_test.py

```diff
@@ -766,3 +766,146 @@ def build(self):
         self.assertIn('Library hello1 found', client.out)
         self.assertIn('hello1 world!', client.out)
         self.assertIn('hello2 world!', client.out)
+
+    def test_same_name_global_target_collision(self):
+        conanfile_tpl = textwrap.dedent("""
+            import os
+            from conans import ConanFile, CMake
+
+            class Conan(ConanFile):
+                name = "{name}"
+                version = "1.0"
+                settings = "os", "compiler", "build_type", "arch"
+                generators = "cmake"
+                exports_sources = "src/*"
+
+                def build(self):
+                    cmake = CMake(self)
+                    cmake.configure(source_folder="src")
+                    cmake.build()
+
+                def package(self):
+                    self.copy("*.h", dst="include", src="src")
+                    self.copy("*.lib", dst="lib", keep_path=False)
+                    self.copy("*.a", dst="lib", keep_path=False)
+
+                def package_info(self):
+                    self.cpp_info.names["cmake_find_package"] = "nonstd"
+                    self.cpp_info.filenames["cmake_find_package"] = "{name}"
+                    self.cpp_info.components["1"].names["cmake_find_package"] = "{name}"
+                    self.cpp_info.components["1"].libs = ["{name}"]
+            """)
+        client = TestClient()
+        for name in ["expected", "variant"]:
+            client.run("new {name}/1.0 -s".format(name=name))
+            client.save({"conanfile.py": conanfile_tpl.format(name=name)})
+            client.run("create . {name}/1.0@".format(name=name))
+        middle_cmakelists = textwrap.dedent("""
+            project(middle)
+            cmake_minimum_required(VERSION 3.1)
+
+            include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+            conan_basic_setup()
+
+            find_package(expected)
+            find_package(variant)
+
+            add_library(middle middle.cpp)
+            target_link_libraries(middle nonstd::expected nonstd::variant)
+            """)
+        middle_h = textwrap.dedent("""
```

Comment:
Use the ``gen_function_h()`` and ``gen_function_cpp()`` test helpers.

Reply:
used only `gen_function_h()` as the cpp file if different